### PR TITLE
[SPARK-22750][SQL] Reuse mutable states when possible

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

The PR introduces a new method `addImmutableStateIfNotExists ` to `CodeGenerator` to allow reusing and sharing the same global variable between different Expressions. This helps reducing the number of global variables needed, which is important to limit the impact on the constant pool.

## How was this patch tested?

added UTs

### CODE REVIEW

**Comment by viirya** [comment]
A high level question is, do we need to share mutable status, if we can compact global variables into array later?

Will sharing mutable status increase the difficulty of debugging codegen in the future?

**Comment by mgaido91** [comment]
@viirya we have seen that using arrays affects performance. Thus if we can reduce their usage it is better.

I don't think that debugging is harder. These variables I made shared are never assigned, but in the initialization. Do you have an other opinion? Or are you thinking for something specific?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -184,6 +190,27 @@ class CodegenContext {
...
+   * Internally, this method calls `addMutableState`.
+   *
+   * @param javaType Java type of the field.
+   * @param variableName Name of the field.
+   * @param initCode The statement(s) to put into the init() method to initialize this field.
+   */
+  def addSingleMutableState(
+      javaType: String,
+      variableName: String,
+      initCode: String = ""): Unit = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: How can we support different initCode for the same name?

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: It is not supported.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 2]
```diff
@@ -184,6 +190,27 @@ class CodegenContext {
...
+    if (!singleMutableStates.contains(variableName)) {
+      addMutableState(javaType, variableName, initCode)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: shall we add an assert here to make sure `initCode` is same with the previous one?

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: if you want, I can add it.

**Comment by viirya** [comment]
@mgaido91  Do you mean the shared global variables are required to be only assigned once (initialization) and never changed again?

**Comment by mgaido91** [comment]
@viirya this is the requirement I followed in this change which ensures that is it safe to share the variable across all the operators, since all the access are read only and there cannot be influences. Maybe this might be relaxed in the future, but if we follow this requirement, we are sure that this is safe.

**Comment by viirya** [comment]
Shall we make them as `final` variables to guarantee this? I think this is an important requirement to prevent something wrong when wrongly using the shared global variables.

**Comment by viirya** [comment]
Oh, the initialization is not right away in declaration.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Please also check if the java type is the same. If one expression uses the same name with different type, we should alert it early.

**Comment by kiszk** [comment]
I have one question. We are implementing #19811 to compact mutable states. Even when it will be merged, does this PR can reduce large number of constant pool entries?

**Comment by mgaido91** [comment]
@kiszk for instance it can remove one entry for every timestamp function (to_timestamp or from_utc_timestamp). Of course #19811 is the most important PR, because it solves the problem. But I think we all agree that if we can avoid to waste global variables it is better and there have been many PRs to avoid the usage of global variables. This is one of the many.

**Comment by mgaido91** [comment]
Jenkins, retest this please

**Comment by kiszk** [comment]
Got it. It seems to be . I have two questions.

1. Is this mutable? It looks it is global, but immutable. IIUC, is there any other better API name?
2. As @viirya pointed out [here](https://github.com/apache/spark/pull/19940#issuecomment-350724536), why don't we declare `final`?

**Comment by mgaido91** [comment]
1. Yes, to ensure that there are no problems, variables' value should not be changed. I agree that we can find a better API name, to enforce this: what about `addImmutableStateIfNotExists`? Or do you have any suggestion?
2. Beacause as @viirya pointed out [here](https://github.com/apache/spark/pull/19940#issuecomment-350724795), they are not initialized in the declaration, but in the `init` method of the generated class, thus it cannot be made final unfortunately.

**Comment by kiszk** [comment]
For 2., I noticed there are two types of initialization. One is in `init` method by using `addPartitionInitializationStatement`. The other is in constructor that includes an initialization statement in `addSingleMutableState`. Can we declare `final` for latter cases?

**Comment by mgaido91** [comment]
I don't think so, it would be very risky, since `splitExpressions` might put the initialization outside the constructor.
What about 1? Any thought?

**Comment by mgaido91** [comment]
@cloud-fan @kiszk @viirya  any more comments on this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 3]
```diff
@@ -170,6 +170,14 @@ class CodegenContext {
   val mutableStates: mutable.ArrayBuffer[(String, String, String)] =
     mutable.ArrayBuffer.empty[(String, String, String)]
 
+  /**
+   * A map containing the mutable states which have been defined so far using
+   * `addSingleMutableState`. Each entry contains the name of the mutable state as key and its
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Is `addSingleMutableState` old one?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala [hunk 4]
```diff
@@ -401,4 +401,16 @@ class CodeGenerationSuite extends SparkFunSuite with ExpressionEvalHelper {
     ctx.addReferenceObj("foo", foo)
     assert(ctx.mutableStates.isEmpty)
   }
+
+  test("SPARK-22750: addSingleMutableState") {
```
Comment (kiszk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala]: Is `addSingleMutableState` old one?

Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala]: yes, thanks, good catch!

**Comment by mgaido91** [comment]
the test failure is not related to the PR. It looks the same R failure we had some days ago, I thought it was solved:
```
checking CRAN incoming feasibility ...Error in .check_package_CRAN_incoming(pkgdir) : 
  dims [product 22] do not match the length of object [0]
```

**Comment by mgaido91** [comment]
Jenkins, retest this please

**Comment by mgaido91** [comment]
Jenkins, retest this please

**Comment by mgaido91** [comment]
the test errors are unrelated to this change. Any other comments @cloud-fan @kiszk @viirya ?

**Comment by mgaido91** [comment]
Jenkins, retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 5]
```diff
@@ -195,6 +195,14 @@ class CodegenContext {
 
   }
 
+  /**
+   * A map containing the mutable states which have been defined so far using
+   * `addImmutableStateIfNotExists`. Each entry contains the name of the mutable state as key and
+   * its Java type and init code as value.
+   */
+  val singleMutableStates: mutable.Map[String, (String, String)] =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: immutableStates?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 6]
```diff
@@ -1193,7 +1196,8 @@ case class ToUTCTimestamp(left: Expression, right: Expression)
         val dtu = DateTimeUtils.getClass.getName.stripSuffix("$")
         val tzTerm = ctx.addMutableState(tzClass, "tz",
           v => s"""$v = $dtu.getTimeZone("$tz");""")
-        val utcTerm = ctx.addMutableState(tzClass, "utc",
+        val utcTerm = "tzUTC"
+        ctx.addImmutableStateIfNotExists(tzClass, utcTerm,
           v => s"""$v = $dtu.getTimeZone("UTC");""")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: unrelated question: in the codebase sometimes we use UTC sometimes we use GMT, is it corrected?

**Comment by cloud-fan** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala (hunk 6 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: yes, there is no difference between them in practice. But I think that being consistent would be better for readability

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**